File: frontend-admin-dashboard/src/routes/study-library/doubt-management/-types/doubt-activity.ts
export type DoubtActivityAction =
    | 'CREATED'
    | 'ASSIGNED'
    | 'UNASSIGNED'
    | 'STATUS_CHANGED'
    | 'REMARK';
export type DoubtActivityActorType = 'USER' | 'RULE' | 'SYSTEM';

/** One audit-trail row from GET /doubts/{id}/activity. Staff-only; learners never receive these. */
export interface DoubtActivity {
    id: string;
    doubt_id: string;
    action: DoubtActivityAction;
    actor_type: DoubtActivityActorType;
    /** Who did it (USER rows). */
    actor_user_id?: string | null;
    /** Who was (un)assigned. */
    target_user_id?: string | null;
    /** Previous workflow status key (STATUS_CHANGED). */
    from_value?: string | null;
    /** New status key, or the status a remark was left on. */
    to_value?: string | null;
    /**
     * Routing rule for RULE rows: `TYPE:<type>:<source>[:<role>]`, `DEFAULT:<source>`, `SUB_ORG`,
     * `ADMIN_FALLBACK`; `DEFAULT_EXCLUDED` on an UNASSIGNED row means a default pill was removed.
     */
    rule_source?: string | null;
    remark?: string | null;
    created_at: string;
}

File: admin_core_service/src/main/java/vacademy/io/admin_core_service/features/doubts/entity/DoubtActivity.java
package vacademy.io.admin_core_service.features.doubts.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.UuidGenerator;

import java.util.Date;

/**
 * One audit-trail row for a doubt: who assigned / un-assigned whom, who changed the status (and to
 * what), and remarks left along the way. {@code actorType=RULE} rows are the auto-assignments made
 * by the institute's routing config at creation time, with {@link #ruleSource} naming the rule.
 */
@Entity
@Table(name = "doubt_activity")
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class DoubtActivity {
    @Id
    @UuidGenerator
    @Column(name = "id")
    private String id;

    @Column(name = "doubt_id")
    private String doubtId;

    /** {@link vacademy.io.admin_core_service.features.doubts.enums.DoubtActivityActionEnum#name()} */
    @Column(name = "action")
    private String action;

    /** {@link vacademy.io.admin_core_service.features.doubts.enums.DoubtActivityActorTypeEnum#name()} */
    @Column(name = "actor_type")
    private String actorType;

    @Column(name = "actor_user_id")
    private String actorUserId;

    @Column(name = "target_user_id")
    private String targetUserId;

    @Column(name = "from_value")
    private String fromValue;

    @Column(name = "to_value")
    private String toValue;

    @Column(name = "rule_source")
    private String ruleSource;

    @Column(name = "remark")
    private String remark;

    @Column(name = "created_at", insertable = false, updatable = false)
    private Date createdAt;
}
